In [ ]:
import pandas as pd
import numpy as np

In [ ]:
df=pd.read_csv("/content/banknotes.csv")
df.head()

conterfeit  Length   Left  Right  Bottom   Top  Diagonal
0           0   214.8  131.0  131.1     9.0   9.7     141.0
1           0   214.6  129.7  129.7     8.1   9.5     141.7
2           0   214.8  129.7  129.7     8.7   9.6     142.2
3           0   214.8  129.7  129.6     7.5  10.4     142.0
4           0   215.0  129.6  129.7    10.4   7.7     141.8

In [ ]:
df['conterfeit'].value_counts()

1    100
0    100
Name: conterfeit, dtype: int64

In [ ]:
X=df.drop(['conterfeit'],axis=1)
y=df['conterfeit']
X.head()

Length   Left  Right  Bottom   Top  Diagonal
0   214.8  131.0  131.1     9.0   9.7     141.0
1   214.6  129.7  129.7     8.1   9.5     141.7
2   214.8  129.7  129.7     8.7   9.6     142.2
3   214.8  129.7  129.6     7.5  10.4     142.0
4   215.0  129.6  129.7    10.4   7.7     141.8

In [ ]:
df.isnull().sum()

conterfeit    0
Length        0
Left          0
Right         0
Bottom        0
Top           0
Diagonal      0
dtype: int64

In [ ]:
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.2,random_state=1)

**Logistic** **Regression**

In [ ]:
from sklearn.linear_model import LogisticRegression
lr=LogisticRegression()
lr.fit(X_train,y_train)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='auto', n_jobs=None, penalty='l2',
                   random_state=None, solver='lbfgs', tol=0.0001, verbose=0,
                   warm_start=False)

In [ ]:
y_pred=lr.predict(X_test)
print(y_pred)

[0 0 0 1 1 1 0 0 0 1 1 0 0 0 1 1 1 0 1 0 0 0 0 0 0 1 0 0 1 0 0 1 1 1 1 0 1
 0 1 1]


In [ ]:
print("Accuracy",(lr.score(X_test,y_test)))

Accuracy 0.975


In [ ]:
from sklearn.metrics import confusion_matrix,classification_report
cm=confusion_matrix(y_test,y_pred)
print(cm)

[[22  1]
 [ 0 17]]


In [ ]:
print(classification_report(y_test,y_pred))

              precision    recall  f1-score   support

           0       1.00      0.96      0.98        23
           1       0.94      1.00      0.97        17

    accuracy                           0.97        40
   macro avg       0.97      0.98      0.97        40
weighted avg       0.98      0.97      0.98        40



First, we make use of the Logistic Regression for our given dataset without any modifications. As we can see from the accuracy of 97.5%, simply performing Logistic Regression is enough for this data set as this gives a significant result but let us explore other options.

**Logistic Regression with Gradient Descent**

In [ ]:
from xgboost import XGBClassifier

model = XGBClassifier()

model.fit(X_train,y_train)

XGBClassifier(base_score=0.5, booster='gbtree', colsample_bylevel=1,
              colsample_bynode=1, colsample_bytree=1, gamma=0,
              learning_rate=0.1, max_delta_step=0, max_depth=3,
              min_child_weight=1, missing=None, n_estimators=100, n_jobs=1,
              nthread=None, objective='binary:logistic', random_state=0,
              reg_alpha=0, reg_lambda=1, scale_pos_weight=1, seed=None,
              silent=None, subsample=1, verbosity=1)

In [ ]:
y_pred = model.predict(X_test)
print(classification_report(y_test,y_pred))

              precision    recall  f1-score   support

           0       1.00      0.96      0.98        23
           1       0.94      1.00      0.97        17

    accuracy                           0.97        40
   macro avg       0.97      0.98      0.97        40
weighted avg       0.98      0.97      0.98        40



In [ ]:
print(confusion_matrix(y_test, y_pred))

[[22  1]
 [ 0 17]]


Next, we use XGBoost model because it uses a gradient descent framework. As we can see from the accuracy of 97.5% there is not any improvement in the model because this is a simple dataset.

**Logistic Regression with LDA**

In [ ]:
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis as LDA
lda = LDA(n_components=2)
X_train = lda.fit_transform(X_train, y_train)
X_test = lda.transform(X_test)


/usr/local/lib/python3.7/dist-packages/sklearn/discriminant_analysis.py:463: ChangedBehaviorWarning: n_components cannot be larger than min(n_features, n_classes - 1). Using min(n_features, n_classes - 1) = min(1, 2 - 1) = 1 components.
  ChangedBehaviorWarning)
/usr/local/lib/python3.7/dist-packages/sklearn/discriminant_analysis.py:469: FutureWarning: In version 0.23, setting n_components > min(n_features, n_classes - 1) will raise a ValueError. You should set n_components to None (default), or a value smaller or equal to min(n_features, n_classes - 1).
  warnings.warn(future_msg, FutureWarning)


In [ ]:
from sklearn.linear_model import LogisticRegression

model1 = LogisticRegression()

model1.fit(X_train,y_train)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='auto', n_jobs=None, penalty='l2',
                   random_state=None, solver='lbfgs', tol=0.0001, verbose=0,
                   warm_start=False)

In [ ]:
y_pred = model1.predict(X_test)
confusion_matrix(y_pred,y_test)

array([[22,  0],
       [ 1, 17]])

In [ ]:
from sklearn.metrics import accuracy_score
accuracy_score(y_test,y_pred)

0.975

Next, we try Feature Space Transformation for our features and then implement it in Logistic Regression. We use Linear Discriminant Analysis to change our features' space to two components and then implementing it with normal Logistic Regression. But the accuracy of 97.5 indicates that there is no significant change in our prediction. Hence, using simple Logistic Regression is ideal because it is easier to implement and runs faster than the other models.